### ISSUE TITLE
Waybar Crash (Fedora 34)

### ISSUE THREAD

**Issue by sclel016** [Alexays/Waybar#1065]
Hi,

The problem signature is reproducible with the Waybar 0.9.5-4 available through DNF and using HEAD on master (97e4b53).

I'm running latest SwayWM on Fedora 34. In a dual monitor setup, Waybar crashes when one of the displays in disabled. I've attached the debug logs from a crash where I disable my laptops internal display while connected to an external monitor.

I'm happy to help in any way possible, love the project!

Stew
[waybar.log](https://github.com/Alexays/Waybar/files/6282473/waybar.log)

**Comment by martinpitt** [comment]
I get that crash on either changing displays (like docking/undocking Thunderbolt) or suspend/resume. When starting wayland in a terminal:
```
(waybar:97366): dbind-WARNING **: 10:28:56.027: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.NameHasNoOwner: Could not activate remote peer.
[2021-04-11 10:28:56.047] [info] Resources files: /home/martin/.config/waybar/config, /etc/xdg/waybar/style.css
[2021-04-11 10:28:56.055] [error] media stopped unexpectedly, is it endless?
[2021-04-11 10:28:56.055] [warning] module battery#bat2: Disabling module "battery#bat2", No battery named BAT2
[2021-04-11 10:28:56.055] [warning] For a functional tray you must have libappindicator-* installed and export XDG_CURRENT_DESKTOP=Unity

** (waybar:97366): WARNING **: 10:28:56.060: gtk-layer-shell v0.6.0 may not work on GTK v3.24.28. If you experience crashes, check https://github.com/wmww/gtk-layer-shell/blob/master/compatibility.md
[2021-04-11 10:28:56.204] [info] Bar configured (width: 1920, height: 17) for output: eDP-1
```

then close / reopen lid:
```
(waybar:97366): GLib-GIO-WARNING **: 10:29:12.512: Invalid id 2 passed to g_bus_unwatch_name()
terminate called without an active exception
Aborted (core dumped)
```

**Comment by martinpitt** [comment]
Ignore me please -- "my" crash is fixed in current master (97e4b53cf3b as well), and but suspending and changing displays work well. Time for another release? There are dozens of commits since 0.9.5

Thank you!

**Comment by uaftab** [comment]
Hi

I am also facing the same crash as reported by the original poster. Waybar 0.9.5-4 available through DNF in SwayWM on Fedora 34. 

When connected to an external display, my crash with the same signature occurs when the internal display of the laptop is disabled.

**Comment by Alexays** [comment]
I'll make another release tomorrow!

**Comment by sclel016** [comment]
Hi Alex,

I'm not sure this issue is can be closed yet. I'm still seeing the crash with the latest from master. I believe @martinpitt was experiencing a different issue that was resolved using latest. Do my original debug logs help you understand the crash?

**Comment by alebastr** [comment]
I've seen that crash on retrace, but didn't have time to take a look until now.

Something in the Fedora's gcc 11 + glibc 2.33 toolchain has finally broken `pthread_cancel` :disappointed:
I know that `pthread_cancel` is evil and there's at least [an issue with `std::thread` destructor](https://gcc.gnu.org/legacy-ml/gcc-help/2015-08/msg00040.html), but the issue with cancellation of `std::condition_variable::wait_for` is a new for me.

Reproducer (with [sleeper_thread.hpp](https://github.com/Alexays/Waybar/blob/97e4b53cf3bbdaa71d9e180b1a50e987cfc7bb43/include/util/sleeper_thread.hpp) from waybar code):

```c++
#include "include/util/sleeper_thread.hpp"

int main(int, char**) {
  using namespace std::chrono_literals;
  using namespace waybar::util;

  SleeperThread thread1;
  thread1 = [&thread1] { thread1.sleep_for(1s); };

  SleeperThread thread2;
  thread2 = [&thread2] { thread2.sleep_for(1s); };

  std::this_thread::sleep_for(1500ms);
  return 0;
}
```
<details>
<summary>backtrace with symbols:</summary>

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
#1  0x00007ffff7a5e8a4 in __GI_abort () at abort.c:79
#2  0x00007ffff7e07a46 in __gnu_cxx::__verbose_terminate_handler () at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007ffff7e1328c in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:48
#4  0x00007ffff7e132f7 in std::terminate () at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:58
#5  0x00007ffff7e12c4c in __cxxabiv1::__gxx_personality_v0 (version=<optimized out>, actions=10, exception_class=0, ue_header=0x7ffff7a32cb0, context=<optimized out>)
    at ../../../../libstdc++-v3/libsupc++/eh_personality.cc:673
#6  0x00007ffff7c18604 in _Unwind_ForcedUnwind_Phase2 () from /lib64/libgcc_s.so.1
#7  0x00007ffff7c18cf2 in _Unwind_ForcedUnwind () from /lib64/libgcc_s.so.1
#8  0x00007ffff7f97506 in __GI___pthread_unwind (buf=<optimized out>) at unwind.c:131
#9  0x00007ffff7f8cc49 in __do_cancel () at ./pthreadP.h:307
#10 sigcancel_handler (sig=32, si=<optimized out>, ctx=<optimized out>) at nptl-init.c:161
#11 sigcancel_handler (sig=<optimized out>, si=<optimized out>, ctx=<optimized out>) at nptl-init.c:126
#12 <signal handler called>
#13 0x00007ffff7f9aa6a in __futex_abstimed_wait_common64 (futex_word=futex_word@entry=0x7fffffffd830, expected=expected@entry=0, clockid=clockid@entry=1,
    abstime=abstime@entry=0x7ffff7a31b60, private=private@entry=0, cancel=cancel@entry=true) at ../sysdeps/nptl/futex-internal.c:74
#14 0x00007ffff7f9aacf in __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x7fffffffd830, expected=expected@entry=0, clockid=clockid@entry=1,
    abstime=abstime@entry=0x7ffff7a31b60, private=private@entry=0) at ../sysdeps/nptl/futex-internal.c:123
#15 0x00007ffff7f94882 in __pthread_cond_wait_common (abstime=0x7ffff7a31b60, clockid=1, mutex=0x7fffffffd838, cond=0x7fffffffd808) at pthread_cond_wait.c:504
#16 __pthread_cond_clockwait (abstime=0x7ffff7a31b60, clockid=1, mutex=0x7fffffffd838, cond=0x7fffffffd808) at pthread_cond_wait.c:672
#17 __pthread_cond_clockwait (cond=0x7fffffffd808, mutex=0x7fffffffd838, clockid=1, abstime=0x7ffff7a31b60) at pthread_cond_wait.c:660
#18 0x0000000000401d17 in std::__condvar::wait_until(std::mutex&, int, timespec&) ()
#19 0x0000000000402e9f in std::cv_status std::condition_variable::__wait_until_impl<std::chrono::duration<long, std::ratio<1l, 1000000000l> > >(std::unique_lock<std::mutex>&, std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > > const&) ()
#20 0x0000000000402c4f in std::cv_status std::condition_variable::wait_until<std::chrono::duration<long, std::ratio<1l, 1000000000l> > >(std::unique_lock<std::mutex>&, std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > > const&) ()
#21 0x000000000040296f in bool std::condition_variable::wait_until<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> >, waybar::util::SleeperThread::sleep_for(std::chrono::duration<long, std::ratio<1l, 1000000000l> >)::{lambda()#1}>(std::unique_lock<std::mutex>&, std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > > const&, waybar::util::SleeperThread::sleep_for(std::chrono::duration<long, std::ratio<1l, 1000000000l> >)::{lambda()#1}) ()
#22 0x000000000040251b in bool std::condition_variable::wait_for<long, std::ratio<1l, 1000000000l>, waybar::util::SleeperThread::sleep_for(std::chrono::duration<long, std::ratio<1l, 1000000000l> >)::{lambda()#1}>(std::unique_lock<std::mutex>&, std::chrono::duration<long, std::ratio<1l, 1000000000l> > const&, waybar::util::SleeperThread::sleep_for(std::chrono::duration<long, std::ratio<1l, 1000000000l> >)::{lambda()#1}) ()
#23 0x00000000004020a8 in waybar::util::SleeperThread::sleep_for(std::chrono::duration<long, std::ratio<1l, 1000000000l> >) ()
#24 0x00000000004012c3 in main::{lambda()#1}::operator()() const ()
#25 0x0000000000401a28 in void std::__invoke_impl<void, main::{lambda()#1}&>(std::__invoke_other, main::{lambda()#1}&) ()
#26 0x00000000004017d6 in std::enable_if<is_invocable_r_v<void, main::{lambda()#1}&>, void>::type std::__invoke_r<void, main::{lambda()#1}&>(main::{lambda()#1}&) ()
#27 0x00000000004015e2 in std::_Function_handler<void (), main::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
#28 0x00000000004022f2 in std::function<void ()>::operator()() const ()
#29 0x0000000000401f65 in waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}::operator()() const ()
#30 0x0000000000403376 in void std::__invoke_impl<void, waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}>(std::__invoke_other, waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}&&) ()
#31 0x000000000040332b in std::__invoke_result<waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}>::type std::__invoke<waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}>(waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}&&) ()
#32 0x00000000004032d8 in void std::thread::_Invoker<std::tuple<waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) ()
#33 0x00000000004032ac in std::thread::_Invoker<std::tuple<waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}> >::operator()() ()
#34 0x0000000000403290 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<waybar::util::SleeperThread::operator=(std::function<void ()>)::{lambda()#1}> > >::_M_run() ()
#35 0x00007ffff7e3fc94 in std::execute_native_thread_routine (__p=0x41aeb0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:82
#36 0x00007ffff7f8e299 in start_thread (arg=0x7ffff7a32640) at pthread_create.c:473
#37 0x00007ffff7b386a3 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
</details>

I don't really have a good idea for addressing that. Here's a list of not-so-good ideas:

 * One possible option is to ditch `std::thread` completely and reimplement everything with direct use of `pthread_*`  (with certain amounts of suffering and without fancy lambdas).
 * Another idea is to split periodic and I/O tasks and use [`std::jthread`](https://en.cppreference.com/w/cpp/thread/jthread) with `request_stop` method for all periodic tasks when compiled with `-Dcpp_std=2a`.
    Note that [`stop_token`](https://en.cppreference.com/w/cpp/thread/stop_token) is not an equivalent of `pthread_cancel` and will not cause interruption of `poll` or any other functions waiting on IO.
 * Third idea is to stop using threads for periodic tasks and migrate everything to [`Glib::SignalTimeout`](https://developer.gnome.org/glibmm/stable/classGlib_1_1SignalTimeout.html). Precision may suffer.

**Comment by danielrparks** [comment]
This issue also affects Arch Linux; it happens every time I unplug my external monitor.
I don't envy whoever has to try to fix that. Threading problems are gross.

**Comment by alebastr** [comment]
I tried to solve this without a major rework in #1132. Please, test and verify that the patch fixes the issue.
Fedora users can use prebuilt package from https://copr.fedorainfracloud.org/coprs/build/2251643

**Comment by sclel016** [comment]
I installed your copr package. Waybar no longer crashes when disabling the display using `swaymsg`! I'm still seeing a crash if I unplug the external display, but this is a great improvement.

**Comment by danielrparks** [comment]
> I installed your copr package. Waybar no longer crashes when disabling the display using `swaymsg`! I'm still seeing a crash if I unplug the external display, but this is a great improvement.

That's odd, it doesn't crash when I unplug my display.

**Comment by alebastr** [comment]
> I'm still seeing a crash if I unplug the external display, but this is a great improvement.

Can you please install `waybar-debuginfo` and `waybar-debugsource` from the same copr and get output of `coredumpctl info -1 /usr/bin/waybar`?

**Comment by sclel016** [comment]
Turns out I had an old version of waybar installed from source that was being interfering with calls to your copr version. After cleaning that up everything works as expected. Thanks for the fix!

**Comment by alebastr** [comment]
Thanks for verifying!

**Comment by PopeRigby** [comment]
Is this bugfix in the latest release?

**Comment by sclel016** [comment]
I'm running 0.9.7-2 from the Fedora 34 updates repo and the issue is gone.

**Comment by PopeRigby** [comment]
Weird, I'm still getting it. I'm using `0.9.7-5` on Arch Linux. I've already rebooted a few times.

**Comment by sclel016** [comment]
Interesting, maybe a regression was introduced between `0.9.7-2` and `0.9.7-5`. @danielrparks confirmed the fix worked in Arch, I wonder if that's still the case.

**Comment by PopeRigby** [comment]
Should I open a new issue or something?

**Comment by PopeRigby** [comment]
I'm on HEAD, and it seems to be fixed, so I guess it wasn't in the latest release.

**Comment by cmprmsd** [comment]
I have this crashing issue, when I close the lid on my laptop:

```
[2021-08-23 08:39:56.197] [debug] GTK widget tree:
window#waybar.background..eDP-1:dir(ltr)
  decoration:dir(ltr)
  box.horizontal:dir(ltr)
    box.horizontal.modules-left:dir(ltr)
      widget:dir(ltr)
        box#workspaces.horizontal:dir(ltr)
      widget:dir(ltr)
        label#mode:dir(ltr)
    box.horizontal.modules-center:dir(ltr)
    box.horizontal.modules-right:dir(ltr)
      widget:dir(ltr)
        label#custom-media:dir(ltr)
      widget:dir(ltr)
        label#idle_inhibitor:dir(ltr)
      widget:dir(ltr)
        label#pulseaudio:dir(ltr)
      widget:dir(ltr)
        label#network:dir(ltr)
      widget:dir(ltr)
        label#cpu:dir(ltr)
      widget:dir(ltr)
        label#memory:dir(ltr)
      widget:dir(ltr)
        label#temperature:dir(ltr)
      widget:dir(ltr)
        label#backlight:dir(ltr)
      widget:dir(ltr)
        label#battery:dir(ltr)
      widget:dir(ltr)
        label#clock:dir(ltr)
      widget:dir(ltr)
        box#tray.horizontal:dir(ltr)

[2021-08-23 08:39:56.316] [info] Bar configured (width: 1920, height: 30) for output: eDP-1
[2021-08-23 08:40:00.415] [debug] Output removed: Unknown 0x05E5
terminate called without an active exception
[1]    5945 abort (core dumped)  waybar -l debug
```

The debug log turns out to not include a lot of details though :(
